invoices/invoice_expiry_watcher_test.go

```diff
@@ -203,3 +204,113 @@ func TestInvoiceExpiryWhenAddingMultipleInvoices(t *testing.T) {
 	test.watcher.Stop()
 	test.checkExpectations()
 }
+
+// TestExpiredHodlInv tests expiration of an already-expired hodl invoice
+// which has no htlcs.
+func TestExpiredHodlInv(t *testing.T) {
+	t.Parallel()
+
+	creationDate := testTime.Add(time.Hour * -24)
+	expiry := time.Hour
+
+	test := setupHodlExpiry(
+		t, creationDate, expiry, 0, channeldb.ContractOpen, nil,
+	)
+
+	test.assertCanceled(t, test.hash)
+	test.watcher.Stop()
+}
+
+// TestAcceptedHodlNotExpired tests that hodl invoices which are in an accepted
+// state are not expired once their time-based expiry elapses, using a regular
+// invoice that expires at the same time as a control to ensure that invoices
+// with that timestamp would otherwise be expired.
+func TestAcceptedHodlNotExpired(t *testing.T) {
+	t.Parallel()
+
+	creationDate := testTime
+	expiry := time.Hour
+
+	test := setupHodlExpiry(
+		t, creationDate, expiry, 0, channeldb.ContractAccepted, nil,
+	)
+	defer test.watcher.Stop()
+
+	// Add another invoice that will expire at our expiry time as a control
+	// value.
+	tsExpires := &invoiceExpiryTs{
+		PaymentHash: lntypes.Hash{1, 2, 3},
+		Expiry:      testTime.Add(expiry),
+		Keysend:     true,
+	}
+	test.watcher.AddInvoices(tsExpires)
+
+	test.mockClock.SetTime(testTime.Add(expiry + 1))
+
+	// Assert that only the ts expiry invoice is expired.
+	test.assertCanceled(t, tsExpires.PaymentHash)
```

Comment:
Is there a way we can assert the invoice is not canceled? Or is it implicitly asserted in this `assertCanceled`?

Reply:
Yeah, just implicitly asserted. Couldn't think of a better way to test it.